File: server/internal/clancert/clancert.go
// Package clancert issues the certificate that puts a player's clan tag on
// their name in game.
//
// # Why a certificate and not a lookup
//
// A game server builds the displayed name from GameConnection::getAuthInfo(),
// and it does so inside onConnect, synchronously. The mod used to fill that
// record by asking this server over HTTP while holding the connection open. A
// signed certificate the player carries removes the round trip, the cache and
// the hold: the game server checks it offline, the way it already checks the
// account certificate TribesNext signed.
//
// # Why not the shipped community certificate
//
// TribesNext has this feature, and its verifier is dead: a community
// certificate is verified with a delegated server's key, that server's own
// certificate is signed by a root key compiled into IFC22.dll, and the only
// certificate ever issued has expired. Nobody but its author can renew it.
//
// We do not need that chain. It exists so an arbitrary community server can
// convince an arbitrary game server with nothing configured between them; we
// have one issuer, and every verifier already holds our key because the
// verifier is our mod. What we cannot self-assert -- that a GUID belongs to a
// real account with a real name -- is exactly what the TribesNext account
// certificate already says, and the game server has checked it before ours is
// looked at. So this certificate only has to bind a clan record to a GUID that
// somebody else established.
//
// # The format
//
//	KeyID <TAB> Issued <TAB> Expire <TAB> GUID <TAB> HexBlob <TAB> Sig
//
// Six tab-separated fields on one line. HexBlob is the hex of the auth-info
// record dbproxy.Certificate produces -- hex because the record itself contains
// tabs and newlines, which are the field and record separators of the string it
// has to travel inside.
//
// Sig is raw RSA over a bare SHA-1 of fields 0..4 joined with tabs: no PKCS#1,
// no ASN.1, no hash prefix. That is not a choice, it is what the verifier does.
// The mod checks it with the engine's rsa_mod_exp(sig, e, n) and compares the
// result to sha1sum(preimage) as hex strings, so the number being signed is the
// integer whose hex representation is the digest.
//
// # Lifetime
//
// Short, and it is the only freshness mechanism there is -- no revocation
// exists, so expiry is also how a player who leaves a clan stops wearing its
// tag. It can be aggressive because an expired certificate costs a tag and
// nothing else: the mod applies what verifies and leaves the player alone
// otherwise. Nothing here can stop somebody playing.
package clancert

import (
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha1"
	"crypto/sha256"
	"crypto/x509"
	"encoding/hex"
	"encoding/pem"
	"fmt"
	"math/big"
	"os"
	"strconv"
	"strings"
	"time"
)

// ErrNoKey is what every entry point answers when the server was started
// without a signing key. Callers turn it into a 404: a deployment that has not
// configured this simply does not offer it, and must still start and serve
// everything else.
var ErrNoKey = fmt.Errorf("no clan signing key configured")

// DefaultTTL is how long an issued certificate stays valid.
//
// Half an hour. The client refreshes at half-life, so a player picks up a clan
// change within about fifteen minutes, and a player who left one keeps the tag
// for at most thirty. Both numbers are cosmetic-only, which is what allows them
// to be this short: nothing is disconnected when a certificate lapses.
//
// There is a floor under this that is not ours. The game server compares
// epoch seconds in TorqueScript, which evaluates numbers as 32-bit floats, so
// timestamps this large quantise to multiples of 128 -- measured in game, where
// a 1800-second lifetime reports itself as 1792. Anything under a few minutes
// would be decided by rounding rather than by time.
const DefaultTTL = 30 * time.Minute

// KeyBits is the size Generate produces.
//
// 2048, against the 4096 TribesNext uses for accounts. The verifier is a modexp
// run inside a script callback on the game server's connect path, once per
// joining player, and this halves both that and the bytes the client has to
// chunk across the command channel. It signs a clan tag, not an identity.
const KeyBits = 2048

// Signer issues certificates. A nil *Signer is the "no key configured" state
// and answers ErrNoKey rather than panicking, so callers need no second flag.
type Signer struct {
	keyID int
	ttl   time.Duration
	key   *rsa.PrivateKey
}

// Load reads a PEM private key from a file.
func Load(path string, keyID int, ttl time.Duration) (*Signer, error) {
	if path == "" {
		return nil, ErrNoKey
	}
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("reading the clan signing key: %w", err)
	}
	return Parse(raw, keyID, ttl)
}

// IsPEM reports whether a configured value is key material rather than a path
// to some.
//
// Both are supported because both are how a key actually arrives. A machine you
// administer has a file; a platform hands you an environment variable and no
// filesystem to write to -- App Platform runs this from a distroless image with
// no volume, so a path there would have nothing to point at.
func IsPEM(s string) bool {
	return strings.Contains(s, "-----BEGIN")
}

// Parse reads a PEM private key.
//
// Both PKCS#1 ("RSA PRIVATE KEY", what openssl genrsa writes) and PKCS#8
// ("PRIVATE KEY", what most modern tooling writes) are accepted, because
// whichever one an operator happens to produce is the one they will try.
func Parse(raw []byte, keyID int, ttl time.Duration) (*Signer, error) {
	if keyID < 0 {
		return nil, fmt.Errorf("key id must not be negative")
	}
	if ttl <= 0 {
		ttl = DefaultTTL
	}

	block, _ := pem.Decode(raw)
	if block == nil {
		return nil, fmt.Errorf("the clan signing key is not PEM")
	}

	var key *rsa.PrivateKey
	switch k, err := x509.ParsePKCS1PrivateKey(block.Bytes); {
	case err == nil:
		key = k
	default:
		any, err8 := x509.ParsePKCS8PrivateKey(block.Bytes)
		if err8 != nil {
			return nil, fmt.Errorf("the clan signing key is neither PKCS#1 nor PKCS#8 RSA: %w", err)
		}
		k8, ok := any.(*rsa.PrivateKey)
		if !ok {
			return nil, fmt.Errorf("the clan signing key is not an RSA key")
		}
		key = k8
	}

	if err := key.Validate(); err != nil {
		return nil, fmt.Errorf("the clan signing key is malformed: %w", err)
	}
	// There is no padding scheme here to compensate for a small modulus, and a
	// forged clan certificate is a forged identity as far as the scoreboard is
	// concerned. 1024 is the floor; Generate makes 2048.
	if key.N.BitLen() < 1024 {
		return nil, fmt.Errorf("the clan signing key is only %d bits; use at least 1024", key.N.BitLen())
	}
	return &Signer{keyID: keyID, ttl: ttl, key: key}, nil
}

// FromKey wraps an already-parsed key. Tests use it; Load is the real path.
func FromKey(key *rsa.PrivateKey, keyID int, ttl time.Duration) *Signer {
	if ttl <= 0 {
		ttl = DefaultTTL
	}
	return &Signer{keyID: keyID, ttl: ttl, key: key}
}

// Generate makes a fresh signing key.
func Generate(bits int) (*rsa.PrivateKey, error) {
	if bits <= 0 {
		bits = KeyBits
	}
	return rsa.GenerateKey(rand.Reader, bits)
}

// MarshalPEM encodes a key for -genkey to write out. PKCS#1, because that is
// what an operator inspecting it with openssl will expect to see.
func MarshalPEM(key *rsa.PrivateKey) []byte {
	return pem.EncodeToMemory(&pem.Block{
		Type:  "RSA PRIVATE KEY",
		Bytes: x509.MarshalPKCS1PrivateKey(key),
	})
}

// Sign issues a certificate for one player.
//
// record is the auth-info blob from dbproxy.Certificate, passed through
// unmodified: this function asserts that record, it does not decide it.
func (s *Signer) Sign(guid, record string, now time.Time) (string, error) {
	if s == nil {
		return "", ErrNoKey
	}
	if guid == "" {
		return "", fmt.Errorf("no guid to issue for")
	}
	// A tab or a newline in the GUID would move a field boundary and let one
	// certificate be read as another. Nothing upstream can produce that today;
	// this is here so nothing downstream has to assume it.
	if strings.ContainsAny(guid, "\t\n") {
		return "", fmt.Errorf("guid contains a separator")
	}

	fields := []string{
		strconv.Itoa(s.keyID),
		strconv.FormatInt(now.Unix(), 10),
		strconv.FormatInt(now.Add(s.ttl).Unix(), 10),
		guid,
		hex.EncodeToString([]byte(record)),
	}

	sig := new(big.Int).Exp(digest(strings.Join(fields, "\t")), s.key.D, s.key.N)

	// Fixed width, leading zeros intact. The value is all the verifier reads,
	// but a constant-length signature keeps the certificate a constant size,
	// which is one less thing to be surprised by when it is chunked across the
	// command channel.
	return strings.Join(append(fields, pad(sig, hexWidth(s.key.N))), "\t"), nil
}

// KeyID is the number stamped into field 0, which is how a game server picks
// the public key to check a certificate with.
func (s *Signer) KeyID() int {
	if s == nil {
		return 0
	}
	return s.keyID
}

// PublicHex is the key as the mod carries it: exponent and modulus in
// lower-case hex, no padding scheme and no ASN.1 around them.
func (s *Signer) PublicHex() (e, n string) {
	if s == nil {
		return "", ""
	}
	return fmt.Sprintf("%x", s.key.E), fmt.Sprintf("%x", s.key.N)
}

// ModSettings is the pair of TorqueScript lines a server operator puts in the
// mod, printed by -genkey and at startup.
//
// Handing over the exact lines rather than the numbers is deliberate: the
// alternative is an operator transcribing 512 hex characters into a settings
// file and finding out it was wrong when tags silently do not appear.
func (s *Signer) ModSettings() string {
	if s == nil {
		return ""
	}
	e, n := s.PublicHex()
	return fmt.Sprintf("$TNBS::ClanKeyE[%d] = \"%s\";\n$TNBS::ClanKeyN[%d] = \"%s\";",
		s.keyID, e, s.keyID, n)
}

// Fingerprint identifies the key without printing it: sha256 of the modulus
// hex, matching what auth.Fingerprint does for the account key. Startup logs it
// so an operator can tell which key a deployment is signing with.
func (s *Signer) Fingerprint() string {
	if s == nil {
		return ""
	}
	_, n := s.PublicHex()
	sum := sha256.Sum256([]byte(n))
	return hex.EncodeToString(sum[:])
}

// TTL is the lifetime issued certificates get.
func (s *Signer) TTL() time.Duration {
	if s == nil {
		return 0
	}
	return s.ttl
}

//-----------------------------------------------------------------------------
// The verifier, as the mod performs it
//-----------------------------------------------------------------------------

// Check is the game server's verification, in Go.
//
// It exists to be tested. TNBrowserServer/tnbserver/clancert.cs is the real
// verifier and this mirrors it step for step -- field count, the recovered
// digest compared as 40 lower-case hex characters, expiry, and the GUID the
// account certificate established. If the two ever disagree, one of them is
// wrong and the Go tests are where that becomes visible.
func Check(e, n, cert, guid string, now time.Time) error {
	fields := strings.Split(cert, "\t")
	if len(fields) != 6 {
		return fmt.Errorf("want 6 fields, got %d", len(fields))
	}

	ei, ok := new(big.Int).SetString(strings.TrimSpace(e), 16)
	if !ok {
		return fmt.Errorf("exponent is not hex")
	}
	ni, ok := new(big.Int).SetString(strings.TrimSpace(n), 16)
	if !ok {
		return fmt.Errorf("modulus is not hex")
	}
	sig, ok := new(big.Int).SetString(strings.TrimSpace(fields[5]), 16)
	if !ok {
		return fmt.Errorf("signature is not hex")
	}

	// rsa_mod_exp(sig, e, n), left-padded to 40, against sha1sum(fields 0..4).
	//
	// The padding is not decoration. The shipped account path does it
	// (t2csri/serverSide.cs:114-115) and the shipped DCE path forgets to
	// (serverSideClans.cs:73-78), which would fail one certificate in sixteen --
	// the ones whose digest starts with a zero nibble.
	recovered := pad(new(big.Int).Exp(sig, ei, ni), 40)
	if recovered != pad(digest(strings.Join(fields[:5], "\t")), 40) {
		return fmt.Errorf("signature does not match")
	}

	expire, err := strconv.ParseInt(fields[2], 10, 64)
	if err != nil {
		return fmt.Errorf("expiry is not a number")
	}
	if now.Unix() > expire {
		return fmt.Errorf("certificate expired")
	}

	if fields[3] != guid {
		return fmt.Errorf("certificate is for %q, not %q", fields[3], guid)
	}

	if _, err := hex.DecodeString(fields[4]); err != nil {
		return fmt.Errorf("record is not hex: %w", err)
	}
	return nil
}

// Record decodes the auth-info blob out of a certificate. Verification is
// Check's job; this only reads.
func Record(cert string) (string, error) {
	fields := strings.Split(cert, "\t")
	if len(fields) != 6 {
		return "", fmt.Errorf("want 6 fields, got %d", len(fields))
	}
	raw, err := hex.DecodeString(fields[4])
	if err != nil {
		return "", fmt.Errorf("record is not hex: %w", err)
	}
	return string(raw), nil
}

//-----------------------------------------------------------------------------

// digest is the number that gets signed: the SHA-1 of the preimage, read as the
// integer its hex representation denotes.
func digest(preimage string) *big.Int {
	sum := sha1.Sum([]byte(preimage))
	return new(big.Int).SetBytes(sum[:])
}

// pad renders a value as lower-case hex, at least width characters wide.
func pad(v *big.Int, width int) string {
	return fmt.Sprintf("%0*x", width, v)
}

// hexWidth is how many hex characters a value modulo n occupies.
func hexWidth(n *big.Int) int {
	return (n.BitLen() + 3) / 4
}

File: TNBrowser/tnbrowser/clancert.cs
// TNBrowser -- the clan certificate a player carries into a game
//
// A game server builds the displayed name from GameConnection::getAuthInfo(),
// synchronously, inside onConnect. This is how the clan half of that record
// gets there: the backend signs it, the client carries it, and a game server
// running TNBrowserServer checks the signature offline. No lookup on the
// connect path, and nothing for the game server to hold a player waiting for.
//
// Format, one line, six tab-separated fields:
//
//    KeyID <TAB> Issued <TAB> Expire <TAB> GUID <TAB> HexBlob <TAB> Sig
//
// # It is only ever sent to a server that asks
//
// The certificate lives in $TNB::ClanCert and NOT in
// $T2CSRI::CommunityCertificate, which is the shipped global the client offers
// to every server that pokes it. That distinction is the whole safety property
// of this design.
//
// A stock TribesNext server receiving a community certificate it cannot resolve
// does not shrug: serverSideClans.cs:156 asks the client for the issuing DCE's
// certificate, clientSideClans.cs:31 has none to send and returns silently,
// serverSide.cs:57 then reschedules the challenge every 250ms, and 15 seconds
// later t2csri_expireClient drops the player with "This is a TribesNEXT server.
// You must install the TribesNEXT client to play." -- telling them to install a
// patch they are already running.
//
// So the send is gated on a server asking for it by name. A server without the
// mod never asks, this client never sends, and joining one is exactly as it was
// before this file existed.
//
// # Timers are relative, and deliberately
//
// Freshness is decided from the certificate's own issued/expire fields as a
// duration, never by comparing an expiry against a local clock. The client has
// no reason to trust its own clock against the backend's, and a machine an hour
// out would otherwise either discard every certificate on arrival or keep them
// long past their life. Relative timers cannot be wrong that way.

//-----------------------------------------------------------------------------
// Fetch
//-----------------------------------------------------------------------------

// Get one if we do not already have one.
//
// This is the shell-open path, and it must be cheap to call repeatedly: opening
// EMAIL and then BROWSER should not cost two certificates. TNBClanCertFetch is
// the unconditional form, for the paths that know the one we hold is stale.
function TNBClanCertEnsure()
{
   if ($TNB::ClanCert !$= "")
      return;

   TNBClanCertFetch();
}

// Ask the backend for a fresh certificate.
//
// Silent about everything. There is no failure here worth putting in front of a
// player: a certificate that does not arrive costs a clan tag, and the pane
// they were actually looking at is unaffected.
function TNBClanCertFetch()
{
   if ($TNB::ClanCertPending)
      return;

   // The guard TNBCertEnsure uses, for the same reason: with no TribesNext
   // account there is no session to negotiate, and nothing should try.
   if (TNBSessionGuid() $= "")
      return;

   $TNB::ClanCertPending = 1;
   TNBApiEnqueueOn($TNB::ClanCertURI, "clancert", "", "TNBClanCertLoaded", "", 0);
}

function TNBClanCertLoaded(%ctx, %status, %result)
{
   $TNB::ClanCertPending = "";

   if (%status !$= "ok")
   {
      // A backend with no signing key answers 404, which arrives here as an
      // unreadable response -- indistinguishable from being offline, and it
      // does not matter which it was. Try again later, rarely.
      TNBClanCertRetryLater();
      return;
   }

   // Copied out now: the parsed tree is freed as soon as this returns.
   %cert = TNBJsonStr(%result, "cert");

   if (getFieldCount(%cert) != 6)
   {
      TNBClanCertRetryLater();
      return;
   }

   %issued = getField(%cert, 1);
   %expire = getField(%cert, 2);
   %life = %expire - %issued;

   // A certificate with no life left in it is not worth carrying, and a
   // negative one means the backend is confused. Either way, do not hold it.
   if (%life <= 0)
   {
      TNBClanCertRetryLater();
      return;
   }

   $TNB::ClanCert = %cert;

   TNBClanCertCancelTimers();
   // Refresh at half life, discard at nine tenths. The gap between the two is
   // the margin for a refresh that fails: it gets the rest of the certificate's
   // life to succeed, and if it never does the stale one is dropped before it
   // expires rather than after.
   $TNB::ClanCertRenew = schedule(%life * 500, 0, "TNBClanCertFetch");
   $TNB::ClanCertDrop = schedule(%life * 900, 0, "TNBClanCertDiscard");

   if ($TNB::Debug)
      echo("TNBrowser: clan certificate held, key " @ getField(%cert, 0) @
           ", " @ %life @ "s");

   // A server asked for this while we had nothing, and we are still on it.
   // Handing it over now is what gets a player tagged who joined before their
   // first certificate arrived -- the server rebuilds the name in place.
   //
   // Compared against the connection that asked, so a fetch that outlived the
   // connection is dropped rather than pushed at whatever server we are on now.
   if ($TNB::ClanCertAsked !$= "")
   {
      %asked = $TNB::ClanCertAsked;
      $TNB::ClanCertAsked = "";

      if (isObject(ServerConnection) && %asked == ServerConnection)
         TNBClanCertSend();
   }
}

// Drop what we have.
//
// Sending nothing is the correct degradation and sending something unverifiable
// is not: the game server would refuse it and the player would be untagged
// either way, but only one of those costs a round trip and a log line.
function TNBClanCertDiscard()
{
   $TNB::ClanCert = "";
   TNBClanCertCancelTimers();
   TNBClanCertFetch();
}

// A fetch that failed, to be tried again later.
//
// The renew timer only. The discard timer must survive this, and an earlier
// version of it did not: cancelling both here meant the first failed refresh
// also cancelled the drop that was going to protect the certificate we still
// hold, and nothing ever re-armed it. The client then went on offering an
// expired certificate to every server it joined -- untagged either way, since
// nothing here can refuse a player, but a transfer and a refusal per join for
// no reason, which is exactly what the discard exists to avoid.
function TNBClanCertRetryLater()
{
   if (isEventPending($TNB::ClanCertRenew))
      cancel($TNB::ClanCertRenew);

   $TNB::ClanCertRenew = schedule($TNB::ClanCertRetryMs, 0, "TNBClanCertFetch");
}

function TNBClanCertCancelTimers()
{
   if (isEventPending($TNB::ClanCertRenew))
      cancel($TNB::ClanCertRenew);
   if (isEventPending($TNB::ClanCertDrop))
      cancel($TNB::ClanCertDrop);

   $TNB::ClanCertRenew = "";
   $TNB::ClanCertDrop = "";
}

//-----------------------------------------------------------------------------
// Delivery
//-----------------------------------------------------------------------------

// A game server running TNBrowserServer asking for the certificate.
//
// 200 characters per command, which is the shipped chunk size
// (clientSideClans.cs:35) and set by what a commandToServer argument will
// carry. The server reassembles and is told when the last one has gone.
//
// %version is what the server announced. Unused today and sent anyway, because
// the shipped protocol's one extension point was exactly this and it cost it
// nothing to have.
function clientCmdtnb_wantClanCert(%version)
{
   if ($TNB::ClanCert !$= "")
   {
      $TNB::ClanCertAsked = "";
      TNBClanCertSend();
      return;
   }

   // Nothing to send yet. Remember who asked and fetch one: the answer is
   // seconds away, and TNBClanCertLoaded hands it over when it lands rather
   // than leaving the player untagged for the whole session.
   //
   // The connection object and not a flag. A flag set here and read after a
   // fetch that outlived the connection would push a certificate at whatever
   // server we are on by then -- and if that one does not run this mod,
   // tnb_clanCertChunk does not exist there and every chunk is a console error
   // on somebody else's server.
   $TNB::ClanCertAsked = ServerConnection;
   TNBClanCertFetch();
}

// Hand the certificate to the server we are connected to, 200 characters at a
// time -- the shipped chunk size (clientSideClans.cs:35), set by what a
// commandToServer argument will carry.
function TNBClanCertSend()
{
   if ($TNB::ClanCert $= "" || !isObject(ServerConnection))
      return;

   %cert = $TNB::ClanCert;
   %len = strlen(%cert);
   for (%i = 0; %i < %len; %i += 200)
      commandToServer('tnb_clanCertChunk', getSubStr(%cert, %i, 200));

   commandToServer('tnb_clanCertDone');

   if ($TNB::Debug)
      echo("TNBrowser: sent clan certificate (" @ %len @ " chars)");
}

echo("TNBrowser: clancert.cs loaded");
